curator-framework/src/main/java/org/apache/curator/framework/imps/ReconfigBuilderImpl.java

```diff
@@ -268,7 +269,7 @@ public void processResult(int rc, String path, Object ctx, byte[] bytes, Stat st
                     client.processBackgroundOperation(data, event);
                 }
             };
-            client.getZooKeeper().reconfig(joining, leaving, newMembers, fromConfig, callback, backgrounding.getContext());
+            ((ZooKeeperAdmin)client.getZooKeeper()).reconfigure(joining, leaving, newMembers, fromConfig, callback, backgrounding.getContext());
```

Comment:
Shall we change the return type of `getZooKeeper` to `ZooKeeperAdmin`?

I don't see an instantiate code path of `NonAdminZookeeperFactory` and from 3.5.3 there is always `ZooKeeperAdmin`. From another perspective, what if `getZooKeeper` didn't return `ZooKeeperAdmin` and we reach to this line? It seems the program exits exceptionally.

Reply:
No - via the `ZookeeperFactory` clients can change which ZooKeeper class is returned. In fact, I added `NonAdminZookeeperFactory` that returns the non-admin ZooKeeper. Of course, clients will need to know they can't use the reconfig APIs when this is done.